pkg/operatorsource/configuring.go

```diff
@@ -70,12 +76,74 @@ func (r *configuringReconciler) Reconcile(ctx context.Context, in *marketplace.O
 
 	out = in
 
-	manifests := r.datastore.GetPackageIDsByOperatorSource(in.GetUID())
+	r.logger.Infof("Downloading from [%s]", in.Spec.Endpoint)
+
+	options, err := SetupAppRegistryOptions(r.client, &in.Spec, in.Namespace)
+	if err != nil {
+		nextPhase = phase.GetNextWithMessage(phase.Configuring, err.Error())
+		return
+	}
+
+	registry, err := r.factory.New(options)
+	if err != nil {
+		nextPhase = phase.GetNextWithMessage(phase.Configuring, err.Error())
+		return
+	}
+
+	manifests, err := registry.ListPackages(in.Spec.RegistryNamespace)
+	if err != nil {
+		nextPhase = phase.GetNextWithMessage(phase.Configuring, err.Error())
+		return
+	}
+
+	if len(manifests) == 0 {
+		err = errors.New("The operator source endpoint returned an empty manifest list")
+
+		// Moving it to 'Failed' phase since human intervention is required to
+		// resolve this situation. As soon as the user pushes new operator
+		// manifest(s) registry sync will detect a new release and will trigger
+		// a new reconciliation.
+		nextPhase = phase.GetNextWithMessage(phase.Failed, err.Error())
+		return
+	}
+
+	r.logger.Infof("Downloaded %d manifest(s) from the operator source endpoint", len(manifests))
+
+	// Before we write to the datastore, lets check to see if there are any packages.
+	// If there are not, we are assuming this is a new operator source. We should force
+	// all catalog source configs to compare their versions to what's in the datastore
+	// after we update it.
+	preUpdateDatastorePackageList := r.datastore.GetPackageIDsByOperatorSource(out.GetUID())
+
+	count, err := r.datastore.Write(in, manifests)
+	if err != nil {
+		if count == 0 {
+			// No operator manifest was written, move to Failed phase.
+			nextPhase = phase.GetNextWithMessage(phase.Failed, err.Error())
+			return
+		}
+
+		r.logger.Infof("There were some faulty operator manifest(s), errors - %v", err)
+		err = nil
+	}
+
+	// Now that we have updated the datastore, lets check if the opsrc is new.
+	// If it is, lets for a resync for catalogsourceconfigs.
+	if preUpdateDatastorePackageList == "" {
+		r.logger.Info("New opsrc detected. Refreshing catalogsourceconfigs.")
+		r.refresher.SendRefresh()
+	}
+
+	packages := r.datastore.GetPackageIDsByOperatorSource(out.GetUID())
+	out.Status.Packages = packages
+
+	r.logger.Infof("Successfully stored %d operator manifest(s)", count)
+	r.logger.Info("Download complete, scheduling for configuration")
```

Comment:
Is it possible for us to abstract some of this code into functions? As is this method is really long and hard to read.

Reply:
+1

Reply:
+1